### ISSUE TITLE
Consider putting the admission review on the webhook context

### ISSUE THREAD

**Issue by dprotaso** [knative/pkg#2656]
> The use case is that I need to parse again the AdmissionRequest from Request.Body to perform further checks in the validating/mutating phase.

Would it make more sense for the `AdmissionRequest` go object to be on the context - instead of having the raw request body and parsing it again yourself?

_Originally posted by @dprotaso in https://github.com/knative/pkg/issues/2583#issuecomment-1331507771_

**Comment by elfotografo007** [comment]
Yes it does. It does not only help to prevent us from parsing the request again, but also helps not to consume the Request.Body bytes when we do so.

**Comment by rahulii** [comment]
hey @dprotaso @elfotografo007 , could you give some pointers to where could I start looking into?
thanks!

**Comment by elfotografo007** [comment]
First, you can check this [PR](https://github.com/knative/pkg/pull/2583/files). Reading the HTTP request body `r.Body` will consume the bytes and you are not able to read it again afterwards. The PR reads it and sets a copy again in `r.Body` so it can be read a second time. The purpose of this issue is to make the `AdmissionRequest` go object to be available in the context `cxt` for later use. 

To accomplish this, you may need to create something similar to `ctx = apis.WithHTTPRequest(ctx, r)` like `ctx = apis.WithAdmissionRequest(ctx, r)` to set it and `apis.GetAdmissionRequest(ctx)` to read it afterwards.

I really don't know about Knative internals so I cannot help you any further than this.

**Comment by dprotaso** [comment]
I would also remove `apis.WithHTTPRequest` - since we would want folks to use these new helpers

**Comment by rahulii** [comment]
> First, you can check this [PR](https://github.com/knative/pkg/pull/2583/files). Reading the HTTP request body `r.Body` will consume the bytes and you are not able to read it again afterwards. The PR reads it and sets a copy again in `r.Body` so it can be read a second time. The purpose of this issue is to make the `AdmissionRequest` go object to be available in the context `cxt` for later use.

@elfotografo007 got it .
I gad one naive doubt though - why can't it read for the second time? In the previous PR its mentioned it returns `EOF`? 🤔

**Comment by elfotografo007** [comment]
When you read the bytes, they are consumed and gone forever. I mitigated the issue by using a TeeReader that reads the bytes and copies them back so they can be read again. This has a problem that if you read them again and don't copy them back, they are gone.

**Comment by TusharMohapatra07** [comment]
Is this issue being worked on ? If not, i'd love to look into the issue 
@elfotografo007

**Comment by elfotografo007** [comment]
I am not working on it, so I guess you can take it.
